<a href="https://colab.research.google.com/github/davidofitaly/notes_03_python_in_data_analysis/blob/main/01_the_basis_of_the_numpy_library.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [15]:
import numpy as np

## 1.1 NumPy – Multidimensional Array

### Creating ndarray arrays

#####An `ndarray` (N-dimensional array) is the core data structure in NumPy. It is a fixed-size, homogeneous array that allows for efficient numerical operations. It is created using the `array` function.


####Examples 1.1

In [ ]:
data_1 = [10, 2.5, 0, 9, 5] # Define a Python list with numeric values
array_1 = np.array(data_1)  # Convert the list into a NumPy ndarray

array_1 # Display the NumPy array

array([10. ,  2.5,  0. ,  9. ,  5. ])

In [ ]:
data_2 = [[4,3,2,1], [9, 8, 7, 6]] # Define a Python list with numeric values
array_2 = np.array(data_2)  # Convert the list into a NumPy ndarray

array_2

array([[4, 3, 2, 1],
       [9, 8, 7, 6]])

### `ndim` and `shape` in NumPy

#####In NumPy, `ndim` and `shape` are attributes of an `ndarray` that describe its structure.

- **`ndim`**: Returns the number of dimensions (axes) of the array.
- **`shape`**: Returns a tuple representing the size of the array along each dimension.

####Examples 1.2

In [ ]:
array_1.ndim # Returns the number of dimensions of array_1

1

In [ ]:
array_2.shape # Returns the shape (size along each axis) of array_2

(2, 4)

### `dtype` in NumPy


- `dtype` stands for "data type" and is an attribute in NumPy that specifies the type of elements in an `ndarray`.
- It defines the type of data that the array holds, such as integers, floats, or strings.
- The `dtype` helps in determining the memory consumption and performance of operations.

####Examples 1.3

In [ ]:
array_1.dtype # array_1.dtype will display the data type of the elements in array_1

dtype('float64')

In [ ]:
array_2.dtype # array_2.dtype will show the data type of the elements in array_2

dtype('int64')

### Numpy Array Creation Functions



| Function      | Description                                                             | Example                                                                 |
|---------------|-------------------------------------------------------------------------|-------------------------------------------------------------------------|
| `array`       | Creates an array from a Python list or tuple.                           | `np.array([1, 2, 3])`                                                   |
| `asarray`     | Converts input to an ndarray, but does not copy if input is already ndarray. | `np.asarray([1, 2, 3])`                                                 |
| `arange`      | Returns an array with evenly spaced values within a specified range.     | `np.arange(0, 10, 2)`                                                  |
| `ones`        | Creates an array of ones with a specified shape and data type.          | `np.ones((2, 3))`                                                      |
| `ones_like`   | Returns an array of ones with the same shape and type as a given array. | `np.ones_like(array_1)`                                                |
| `zeros`       | Creates an array of zeros with a specified shape and data type.         | `np.zeros((2, 3))`                                                     |
| `zeros_like`  | Returns an array of zeros with the same shape and type as a given array. | `np.zeros_like(array_1)`                                               |
| `empty`       | Creates an uninitialized array with a specified shape.                  | `np.empty((2, 3))`                                                     |
| `empty_like`  | Returns an uninitialized array with the same shape and type as a given array. | `np.empty_like(array_1)`                                               |
| `full`        | Creates an array filled with a specified value.                          | `np.full((2, 3), 7)`                                                   |
| `full_like`   | Returns an array filled with a specified value, matching the shape and type of a given array. | `np.full_like(array_1, 7)`                                             |
| `eye`         | Creates a 2-D identity matrix with ones on the diagonal.                | `np.eye(3)`                                                            |
| `identity`    | Alias for `eye`. Creates a 2-D identity matrix.                         | `np.identity(3)`                                                       |


####Examples 1.4

In [ ]:
data_3 = [1,2,3,4,5]  # Define a Python list with numeric values
array_3 = np.array(data_3) # Convert the list into a NumPy ndarray

array_3 # Display the NumPy array

array([1, 2, 3, 4, 5])

In [ ]:
asaaray_1 = np.asarray(array_3) # If the input is already a NumPy array, it will return the same object without making a copy.

asaaray_1 # Display the resulting NumPy array

array([1, 2, 3, 4, 5])

In [ ]:
np.arange(7) # It starts from 0 by default and increments by 1.

array([0, 1, 2, 3, 4, 5, 6])

In [ ]:
ones_array = np.ones((4,5))  # Creates a 4x5 array filled with ones.

ones_array # Displays the array

array([[1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1.]])

In [ ]:
ones_like = np.ones_like(ones_array) # Creates an array with the same shape and type as ones_1, filled with ones.

ones_like # Displays the new array

array([[1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1.]])

In [ ]:
zeros_array = np.zeros((2 ,3, 4)) # Creates a 3D array with shape (2, 3, 4), filled with zeros.

zeros_array # Displays the created array

array([[[0., 0., 0., 0.],
        [0., 0., 0., 0.],
        [0., 0., 0., 0.]],

       [[0., 0., 0., 0.],
        [0., 0., 0., 0.],
        [0., 0., 0., 0.]]])

In [ ]:
zeors_like_array = np.zeros_like(zeros_array) # Creates an array with the same shape and type as zeros_array, filled with zeros.

zeors_like_array # Displays the created array

array([[[0., 0., 0., 0.],
        [0., 0., 0., 0.],
        [0., 0., 0., 0.]],

       [[0., 0., 0., 0.],
        [0., 0., 0., 0.],
        [0., 0., 0., 0.]]])

In [ ]:
empty_array = np.empty((2, 3, 2)) # Creates an uninitialized array with the specified shape.

empty_array # Displays the created array

array([[[0., 0.],
        [0., 0.],
        [0., 0.]],

       [[0., 0.],
        [0., 0.],
        [0., 0.]]])

In [ ]:
full_array = np.full((3, 4), 5) # Creates a 3x4 array filled with the value 5.

full_array # Displays the created array


array([[5, 5, 5, 5],
       [5, 5, 5, 5],
       [5, 5, 5, 5]])

In [ ]:
full_like_array = np.full_like(full_array, 7) # Creates an array with the same shape and type as full_array, filled with the value 7.

full_like_array # Displays the created array

array([[7, 7, 7, 7],
       [7, 7, 7, 7],
       [7, 7, 7, 7]])

In [ ]:
eye_array = np.eye(5) # Creates a 5x5 identity matrix (ones on the diagonal, zeros elsewhere)

eye_array # Displays the created identity matrix

array([[1., 0., 0., 0., 0.],
       [0., 1., 0., 0., 0.],
       [0., 0., 1., 0., 0.],
       [0., 0., 0., 1., 0.],
       [0., 0., 0., 0., 1.]])

In [ ]:
identity_array = np.identity(5) # Creates a 5x5 identity matrix using the np.identity() function

identity_array # Displays the created identity matrix

array([[1., 0., 0., 0., 0.],
       [0., 1., 0., 0., 0.],
       [0., 0., 1., 0., 0.],
       [0., 0., 0., 1., 0.],
       [0., 0., 0., 0., 1.]])

### `astype` in NumPy

- `astype()` is a method in NumPy that is used to **cast** or **convert** the data type of an existing array to a specified type.
- It creates a **new array** with the desired data type, leaving the original


In [ ]:
float_array = np.array([1.5, 2.3, 3.7, 4.9]) # Creating a NumPy array of floats

int_array = float_array.astype(int)  # Converts the float array to integers by truncating the decimal part

int_array # Displays the resulting integer array


array([1, 2, 3, 4])

In [ ]:
convert_to_string_array = np.array([1, 2, 3, 4], dtype=str) # Creates a NumPy array with integer elements and converts them to strings

convert_to_string_array # Displays the resulting array with string elements


array(['1', '2', '3', '4'], dtype='<U1')

In [ ]:
convert_to_float_array = np.array([1, 2, 3, 4], dtype=float) # Creates a NumPy array with integer elements and converts them to floats

convert_to_float_array # Displays the resulting array with float elements

array([1., 2., 3., 4.])

### Arithmetic Operations with NumPy Arrays

#####NumPy provides efficient and vectorized arithmetic operations on arrays, allowing you to perform element-wise operations directly on arrays without the need for explicit loops. These operations are typically faster and more efficient compared to traditional Python loops.

##### Basic Arithmetic Operations:
1. **Addition (`+`)**: Adds corresponding elements of two arrays.
2. **Subtraction (`-`)**: Subtracts corresponding elements of one array from another.
3. **Multiplication (`*`)**: Multiplies corresponding elements of two arrays.
4. **Division (`/`)**: Divides corresponding elements of one array by another.
5. **Exponentiation (`^`)**: Raises each element of the array to the power of the corresponding element in another array.
6. **Modulo (`%`)**: Computes the remainder of the division of corresponding elements.

####Examples 1.5

In [ ]:
# Create two arrays
a = np.array([1, 2, 3])
b = np.array([4, 5, 6])

In [ ]:
# Addition
add_result = a + b
print("Addition:", add_result)

Addition: [5 7 9]


In [ ]:
# Subtraction
sub_result = a - b
print("Subtraction:", sub_result)

Subtraction: [-3 -3 -3]


In [ ]:
# Multiplication
mul_result = a * b
print("Multiplication:", mul_result)

Multiplication: [ 4 10 18]


In [ ]:
# Division
div_result = a / b
print("Division:", div_result)

Division: [0.25 0.4  0.5 ]


In [ ]:
# Exponentiation
exp_result = a ** b
print("Exponentiation:", exp_result)

Exponentiation: [  1  32 729]


In [ ]:
# Modulo
mod_result = a % b
print("Modulo:", mod_result)

Modulo: [1 2 3]


### Indexing in NumPy Arrays


#####Indexing in NumPy allows you to access and modify elements in an array using their positions. You can access elements using a single index, multiple indices, or slices. NumPy also supports advanced indexing techniques, including boolean indexing and fancy indexing.


####Examples 1.6



*   Basics of indexing





In [ ]:
array3d = np.array([[[8, 7, 1], [4, 7, 6]], [[7, 5, 9], [2, 11, 0]]])
array3d

array([[[ 8,  7,  1],
        [ 4,  7,  6]],

       [[ 7,  5,  9],
        [ 2, 11,  0]]])

In [ ]:
array3d[0] # Accessing the first 2D slice (row) of the 3D array 'array3d'.

array([[8, 7, 1],
       [4, 7, 6]])

In [ ]:
array3d[1] # Accessing the second 2D slice (row) of the 3D array 'array3d'.

array([[ 7,  5,  9],
       [ 2, 11,  0]])



*   Indexing by means of clippings



In [ ]:
array2d = np.array([[1,2,3], [5,9,1], [2,7,0]])
array2d

array([[1, 2, 3],
       [5, 9, 1],
       [2, 7, 0]])

In [ ]:
array2d[:2, 1:]

array([[2, 3],
       [9, 1]])

In [ ]:
array2d[2, :]

array([2, 7, 0])

In [ ]:
array2d[:, :2]

array([[1, 2],
       [5, 9],
       [2, 7]])

In [ ]:
array2d[2, 2]

0

In [ ]:
array2d[1, :2]

array([5, 9])



*   Indexing and logical values


In [ ]:
names_array = np.array(["John", "Alice", "Bob", "Emma", "James", "Sophia", "John"])
data_array = np.array([[4,7], [8,2], [0,5], [4,8], [2,3], [9,6], [1, 6]])

print(names_array)
print(data_array)

['John' 'Alice' 'Bob' 'Emma' 'James' 'Sophia' 'John']
[[4 7]
 [8 2]
 [0 5]
 [4 8]
 [2 3]
 [9 6]
 [1 6]]


In [ ]:
names_array == "John"

array([ True, False, False, False, False, False,  True])

In [ ]:
data_array[names_array =="Emma"]

array([[4, 8]])

In [ ]:
data_array[names_array == "John"]

array([[4, 7],
       [1, 6]])

In [ ]:
names_array != "Sophia"

array([ True,  True,  True,  True,  True, False,  True])

In [ ]:
data_array[~(names_array == "James")]

array([[4, 7],
       [8, 2],
       [0, 5],
       [4, 8],
       [9, 6],
       [1, 6]])

In [ ]:
mask = (names_array == "Alice") | (names_array == "Bob")
data_array[mask]

array([[8, 2],
       [0, 5]])

In [ ]:
data_array[data_array < 5] = 0
data_array

array([[0, 7],
       [8, 0],
       [0, 5],
       [0, 8],
       [0, 0],
       [9, 6],
       [0, 6]])

In [ ]:
data_array[names_array !="Emma"] = 9
data_array

array([[9, 9],
       [9, 9],
       [9, 9],
       [0, 8],
       [9, 9],
       [9, 9],
       [9, 9]])



*   Fancy indexing



In [ ]:
arr = np.array([[1, 2], [3, 4], [5, 6]]) # Creating a 2D NumPy array with shape (3, 2)

fancy_arr = arr[[1, 0], [0, 1]] # Fancy indexing: Selecting elements at positions (1, 0) and (0, 1)
fancy_arr  # Displays the resulting

array([3, 2])

In [ ]:
arr = np.array([[1, 2, 3], [3, 4, 9], [5, 6, 0]]) # Creating a 2D NumPy array with shape (3, 3)

fancy_arr = arr[[2, 1 ], [0, 2]] # Fancy indexing: Selecting elements at positions (1, 0) and (0, 1)
fancy_arr  # Displays the resulting

array([5, 9])

### Transposing Arrays in NumPy


- **Transpose**: Transposing an array means flipping it over its diagonal, converting rows to columns and vice versa.
- **Function**: You can transpose a NumPy array using `.T` or `np.transpose()`.

####Examples 1.7

In [ ]:
arr = np.array([[1, 2], [3, 4]]) # Creating a 2D array

transposed_arr = np.transpose(arr) # Transposing the array

transposed_arr # Printing the transposed array

array([[1, 3],
       [2, 4]])

In [ ]:
arr = np.arange(15).reshape(3,5) # Create an array of integers from 0 to 14, and reshape it to a 3x5 matrix

arr.T # Transpose the array, swapping its rows and columns

array([[ 0,  5, 10],
       [ 1,  6, 11],
       [ 2,  7, 12],
       [ 3,  8, 13],
       [ 4,  9, 14]])

## 1.2 Generation of pseudo-random numbers

### Generating Pseudorandom Numbers in NumPy

- **Pseudorandom numbers** are numbers generated by an algorithm that mimics randomness. While the numbers appear random, they are actually deterministic, based on a starting value called the **seed**.
  
- **NumPy's `random` module** provides a wide range of functions to generate pseudorandom numbers, such as:
  
  - `np.random.rand()`: Generates numbers from a uniform distribution between 0 and 1.
  - `np.random.randint()`: Generates random integers within a specified range.
  - `np.random.randn()`: Generates random numbers from a standard normal distribution (mean = 0, standard deviation = 1).
  - `np.random.choice()`: Randomly selects elements from an array.

- **Setting the seed**:
  - The `np.random.seed()` function is used to set the seed, allowing you to generate the same sequence of random numbers every time the code is run (useful for reproducibility).

Examples 1.8

In [ ]:
np.random.seed(42) # Set the seed for reproducibility, so the random numbers will be the same each time

random_float = np.random.rand(15).reshape(3,5) # Generate 15 random numbers from a uniform distribution between 0 and 1

print(random_float) # Print the resulting 3x5 matrix of random numbers


[[0.37454012 0.95071431 0.73199394 0.59865848 0.15601864]
 [0.15599452 0.05808361 0.86617615 0.60111501 0.70807258]
 [0.02058449 0.96990985 0.83244264 0.21233911 0.18182497]]


In [ ]:
random_int = np.random.randint(5, size=(1, 11)) # Generate 11 random integers between 0 and 4 (inclusive) and reshape them into a 1x11 array

print(random_int) # Print the resulting array of random integers

[[0 3 1 4 3 0 0 2 2 1 3]]


In [ ]:
random_rand = np.random.rand(10) # Generate an array of 10 random float numbers between 0 and 1 using a uniform distribution

print(random_rand) # Print the resulting array of random floats

[0.17336465 0.39106061 0.18223609 0.75536141 0.42515587 0.20794166
 0.56770033 0.03131329 0.84228477 0.44975413]


## 1.3 Universal functions

### Universal Functions (ufuncs) in NumPy

#####Universal Functions (ufuncs) are functions that operate element-wise on arrays. They are highly efficient and can be classified into two categories based on the number of arguments they accept: **one-argument functions** and **two-argument functions**.

##### 1. One-Argument Functions
#####These functions perform operations on each element of the input array independently.

| Function               | Description                                                       | Example                    |
|------------------------|-------------------------------------------------------------------|----------------------------|
| `np.sqrt`              | Computes the square root of each element.                         | `np.sqrt([4, 9, 16])`      |
| `np.exp`               | Calculates the exponential of each element.                       | `np.exp([1, 2, 3])`        |
| `np.log`               | Calculates the natural logarithm of each element.                 | `np.log([1, 10, 100])`     |
| `np.abs`               | Returns the absolute value of each element.                       | `np.abs([-1, -2, -3])`     |
| `np.sin`               | Computes the sine of each element.                                | `np.sin([0, np.pi/2])`     |

#### 2. Two-Argument Functions
#####These functions perform operations on pairs of elements, one from each of the two input arrays.

| Function               | Description                                                       | Example                        |
|------------------------|-------------------------------------------------------------------|--------------------------------|
| `np.add`               | Adds elements from two arrays element-wise.                       | `np.add([1, 2], [3, 4])`       |
| `np.subtract`          | Subtracts elements of the second array from the first.            | `np.subtract([10, 20], [1, 2])`|
| `np.multiply`          | Multiplies elements from two arrays element-wise.                 | `np.multiply([1, 2], [3, 4])`  |
| `np.divide`            | Divides elements from the first array by those of the second.     | `np.divide([10, 20], [2, 5])`  |
| `np.maximum`           | Returns the element-wise maximum between two arrays.              | `np.maximum([1, 3], [2, 2])`   |

#####These functions are part of **NumPy's ufunc** library and are optimized for performance, allowing for fast, element-wise operations on large arrays.


####Examples 1.9




* one-argument functions



In [26]:
arr = np.arange(17)

arr

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16])

In [25]:
np.sqrt(arr) # Computes the square root of each element

array([[0.        , 1.        , 1.41421356, 1.73205081, 2.        ,
        2.23606798, 2.44948974, 2.64575131, 2.82842712, 3.        ,
        3.16227766, 3.31662479, 3.46410162, 3.60555128, 3.74165739,
        3.87298335, 4.        ]])

In [27]:
np.exp(arr) # Calculates the exponential of each element

array([1.00000000e+00, 2.71828183e+00, 7.38905610e+00, 2.00855369e+01,
       5.45981500e+01, 1.48413159e+02, 4.03428793e+02, 1.09663316e+03,
       2.98095799e+03, 8.10308393e+03, 2.20264658e+04, 5.98741417e+04,
       1.62754791e+05, 4.42413392e+05, 1.20260428e+06, 3.26901737e+06,
       8.88611052e+06])

In [28]:
np.log(arr) # Calculates the natural logarithm of each element.

<ipython-input-28-a67b4ae04e95>:1: RuntimeWarning: divide by zero encountered in log
  np.log(arr)


array([      -inf, 0.        , 0.69314718, 1.09861229, 1.38629436,
       1.60943791, 1.79175947, 1.94591015, 2.07944154, 2.19722458,
       2.30258509, 2.39789527, 2.48490665, 2.56494936, 2.63905733,
       2.7080502 , 2.77258872])



*   two-argument functions


In [31]:
x = np.random.randn(8) # Generate an array of 8 random float numbers between 0 and 1 using a uniform distribution
y = np.random.randn(8) # Generate an array of 8 random float numbers between 0 and 1 using a uniform distribution

print(x)
print(y)

[-1.56306679 -0.79563099  0.9305844   0.67776741  0.69844026  0.17360206
  0.66228451  0.24112216]
[-0.16820509  1.16476865 -0.24574769 -0.76966777  1.21217234  1.2334968
 -1.57415529  0.63795294]


In [32]:
np.add(x,y) # Adds elements from two arrays

array([-1.73127188,  0.36913765,  0.68483671, -0.09190036,  1.9106126 ,
        1.40709887, -0.91187077,  0.87907511])

In [33]:
np.subtract(x,y) # Subtracts elements of the second array from the first

array([-1.39486171, -1.96039964,  1.17633209,  1.44743518, -0.51373208,
       -1.05989474,  2.2364398 , -0.39683078])

In [34]:
np.multiply(x,y) # Multiplies elements from two arrays element-wise

array([ 0.26291579, -0.92672603, -0.22868896, -0.52165573,  0.84662996,
        0.21413759, -1.04253867,  0.15382459])

In [35]:
np.divide(x,y) # Divides elements from the first array by those of the second

array([ 9.29262493, -0.68308071, -3.78674734, -0.88059736,  0.57618891,
        0.14073978, -0.42072375,  0.3779623 ])

In [36]:
np.maximum(x,y) # Returns the element-wise maximum between two arrays

array([-0.16820509,  1.16476865,  0.9305844 ,  0.67776741,  1.21217234,
        1.2334968 ,  0.66228451,  0.63795294])

## 1.4 Programming with arrays

### Logical Conditional Operations in NumPy

#####Logical operations in NumPy allow you to perform element-wise conditional checks on arrays. These operations are useful when filtering or transforming data based on specific conditions.


- **`np.where(condition, x, y)`**:
  - Returns elements from `x` or `y` depending on whether the condition is `True` or `False`. For each element, if the condition is `True`, the element from `x` is chosen, otherwise from `y`.


- **`np.all(condition)`**:
  - Checks if **all** elements of the condition are `True`. Returns `True` if all elements satisfy the condition, otherwise `False`.


- **`np.any(condition)`**:
  - Checks if **any** element of the condition is `True`. Returns `True` if at least one element satisfies the condition, otherwise `False`.


- **`np.count_nonzero(arr)`**:
  - Counts the number of non-zero elements in an array.


####Examples 1.10



*   np.where



In [51]:
arr = np.random.randn(15) # Generate an array of 15 random numbers from a normal distribution (mean = 0, std = 1)

arr # Display the resulting array


array([ 0.35778736,  0.56078453,  1.08305124,  1.05380205, -1.37766937,
       -0.93782504,  0.51503527,  0.51378595,  0.51504769,  3.85273149,
        0.57089051,  1.13556564,  0.95400176,  0.65139125, -0.31526924])

In [57]:
np.where(arr > 0, 2, -2).reshape(3,-1)


array([[ 2,  2,  2,  2, -2],
       [-2,  2,  2,  2,  2],
       [ 2,  2,  2,  2, -2]])

In [56]:
np.where(arr > 0.5, 2, arr)

array([ 0.35778736,  2.        ,  2.        ,  2.        , -1.37766937,
       -0.93782504,  2.        ,  2.        ,  2.        ,  2.        ,
        2.        ,  2.        ,  2.        ,  2.        , -0.31526924])



*   np.all


In [73]:
bools = np.array([True, False, False, True, False, True, True]) # `np.all(bools)` checks if all values in the `bools` array are True

np.all(bools) # Returns `False` because not all elements are True

False

In [75]:
bools = np.array([True, True, 1, True, 2 , True, 0.93]) # `np.all(bools)` checks if all values in the `bools` array are True

np.all(bools) # Returns `True` because all elements are interpreted as True values (non-zero)

True



*   np.any



In [77]:
bools = np.any([True, False, False, True, False, True, True]) # `np.any(bools)` checks if any element in the `bools` array is True

np.all(bools) # The result is `True` because there are True values in the array

True

In [78]:
bools = np.any([True, True, 1, True, 2 , True, 0.93])  # `np.any(bools)` checks if any element in the `bools` array is True

np.all(bools)  # The result is `True` because there are True values in the array

True



*   np.count_nonzero



In [81]:
bools = np.count_nonzero([True, True, False, True, True , True, False]) # `np.count_nonzero()` counts the number of non-zero (or True) values in the input array.

bools # The result of `np.count_nonzero()` will be `5`, as there are 5 True values in the array.

5

### Mathematical and Statistical Methods in NumPy



#####NumPy provides a variety of mathematical and statistical functions that can be used to perform operations on arrays efficiently.

- **`np.mean(array)`**:
  - Calculates the mean (average) of the array elements.


- **`np.sum(array)`**:
  - Computes the sum of the array elements.


- **`np.cumsum(array)`**:
  - Returns the cumulative sum of the elements along a given axis.

- **`np.std(array)`**:
  - Computes the standard deviation of the array elements, which measures the amount of variation.


- **`np.var(array)`**:
  - Calculates the variance of the array elements, which is the square of the standard deviation.


- **`np.min(array)`**:
  - Finds the minimum value in the array.


- **`np.max(array)`**:
  - Finds the maximum value in the array.


- **`np.argmin(array)`**:
  - Returns the index of the minimum value in the array.


- **`np.argmax(array)`**:
  - Returns the index of the maximum value in the array.



####Examples 1.11

In [61]:
x = np.arange(20).reshape(4,-1) # Generates an array with values from 0 to 19.

x # Output will be a 4x5 array:


array([[ 0,  1,  2,  3,  4],
       [ 5,  6,  7,  8,  9],
       [10, 11, 12, 13, 14],
       [15, 16, 17, 18, 19]])

In [62]:
np.mean(x) # Calculates the mean (average) of the array elements


9.5

In [63]:
np.sum(x) # Computes the sum of the array elements

190

In [64]:
np.cumsum(x, axis=0) # Returns the cumulative sum of the elements along a given axis

array([[ 0,  1,  2,  3,  4],
       [ 5,  7,  9, 11, 13],
       [15, 18, 21, 24, 27],
       [30, 34, 38, 42, 46]])

In [65]:
np.cumsum(x, axis=1) # Returns the cumulative sum of the elements along a given axis

array([[ 0,  1,  3,  6, 10],
       [ 5, 11, 18, 26, 35],
       [10, 21, 33, 46, 60],
       [15, 31, 48, 66, 85]])

In [66]:
np.std(x) # Computes the standard deviation of the array elements, which measures the amount of variation

5.766281297335398

In [67]:
np.var(x) # Calculates the variance of the array elements, which is the square of the standard deviation

33.25

In [68]:
np.min(x) #Finds the minimum value in the array

0

In [69]:
np.max(x) #Finds the maximum value in the array

19

In [70]:
np.argmin(x) #Returns the index of the minimum value in the array

0

In [71]:
np.argmax(x) #Returns the index of the maximum value in the array

19

### Sorting in NumPy

#####Sorting is the process of arranging elements in a specified order. In NumPy, several functions allow you to sort arrays efficiently.

##### Functions for Sorting:
1. **`np.sort()`**: Returns a sorted copy of the input array.
    - Syntax: `np.sort(arr)`
    - Example: `np.sort([3, 1, 2])` returns `[1, 2, 3]`.
    
2. **`np.argsort()`**: Returns the indices that would sort an array.
    - Syntax: `np.argsort(arr)`
    - Example: `np.argsort([3, 1, 2])` returns `[1, 2, 0]`.

3. **`np.sort()` with axis**: Allows sorting along a specific axis.
    - Syntax: `np.sort(arr, axis=0)` (Sorts by columns)
    - Syntax: `np.sort(arr, axis=1)` (Sorts by rows)


####Examples 1.12

In [114]:
# arr_1, arr_2, arr_3, arr_4, arr_5, arr_6: Creating six 1x11 arrays of random integers from 0 to 9 using np.random.randint
arr_1 = np.random.randint(10, size=(1,11))
arr_2 = np.random.randint(10, size=(1,11))
arr_3 = np.random.randint(10, size=(1,11))
arr_4 = np.random.randint(10, size=(1,11))
arr_5 = np.random.randint(10, size=(1,11))
arr_6 = np.random.randint(10, size=(1,11))

# Printing each array to visualize their content
print(arr_1)
print(arr_2)
print(arr_3)
print(arr_4)
print(arr_5)
print(arr_6)

[[4 4 9 9 2 0 4 8 0 2 3]]
[[0 0 7 1 7 6 9 9 1 5 5]]
[[2 1 0 5 4 8 0 6 4 4 1]]
[[2 6 5 1 5 1 1 1 2 1 3]]
[[8 5 0 7 6 9 2 0 4 3 9]]
[[7 0 9 0 3 7 4 1 5 4 1]]




*   np.sort



In [115]:
np.sort(arr_1) # Returns a sorted version of arr_1 without modifying the original array

array([[0, 0, 2, 2, 3, 4, 4, 4, 8, 9, 9]])

In [116]:
np.sort(arr_2, axis=0) # Returns a sorted version of arr_2 along the specified axis

array([[0, 0, 7, 1, 7, 6, 9, 9, 1, 5, 5]])

In [117]:
np.sort(arr_3, axis=1) # Returns a sorted version of arr_3 along the specified axis

array([[0, 0, 1, 1, 2, 4, 4, 4, 5, 6, 8]])

In [118]:
np.argsort(arr_4) # Returns the indices of arr_4 that would sort the array

array([[ 3,  5,  6,  7,  9,  0,  8, 10,  2,  4,  1]])

### Unique Values in NumPy

- The function `np.unique()` is used to find the unique elements in an array.
- It returns a sorted array of unique elements from the input array, removing any duplicates.
- By default, it sorts the unique values in ascending order.



In [120]:
arr = np.array([1, 2, 2, 3, 3, 3, 4, 5, 5]) # Creating a NumPy array with some duplicate values

unique_values = np.unique(arr) # Using np.unique to get unique values from the array

unique_values # Displaying the unique values from the array

array([1, 2, 3, 4, 5])